agent/internal/docker.go

```diff
@@ -304,18 +304,175 @@ func (d *dockerActor) sendAuxLog(ctx *actor.Context, msg string) {
+	Backoff time.Time
+}
+
+// renderProgress generates human-readable and log-file-friendly progress messages.
+//
+// Every layer goes through the following stages:
+// - 1 Pulling fs layer (ID but no size)
+// - 1 Waiting (ID but no size)
+// - 1+ Downloading
+// - 1 Verifying Checksum
+// - 1 Download Complete
+// - 1+ Extracting
+// - 1 Pull Complete
+//
+// You can't really estimate global progress because the log stream doesn't tell you how big the
+// full download size is at any point, it only tells you how big each layer is, and only when that
+// layer starts downloading.  The downloads are staggered, so when many layers are present you
+// wouldn't know the full download size until you're basically done.
+//
+// Showing a per-layer status bar is practically impossible without an interactive terminal (as
+// docker run would have).
+//
+// So instead we create a weighted-average status bar, where every layer's download and extraction
+// count as equal parts.  The status bar ends up pretty jerky but it still gives a "sensation" of
+// progress; things don't look frozen, the user has a rough idea of how far along you are, and the
+// logs are still sane afterwards.
+func (f *pullLogFormatter) RenderProgress() string {
+	var downloaded int64
+	var extracted int64
+	progress := 0.0
+	for _, id := range f.Order {
+		info := f.Known[id]
+		downloaded += info.Downloaded
+		extracted += info.Extracted
+		switch {
+		case !info.DownloadStarted:
+			// no progress on this layer
+		case info.Extracted == info.Total:
+			// this layer is complete
+			progress += 1.0
+		case info.Downloaded == info.Total:
+			// download complete, extraction in progress
+			progress += 0.5 + 0.5*float64(info.Extracted)/float64(info.Total)
+		default:
+			progress += 0.5 * float64(info.Extracted) / float64(info.Total)
+		}
+	}
+
+	// normalize by layer count
+	progress /= float64(len(f.Known))
+
+	// 40-character progress bar
+	prog := int(40.0 * progress)
+
+	bar := ""
+	for i := 0; i < 40; i++ {
+		if prog > i {
+			bar += "="
+		} else {
+			bar += " "
+		}
+	}
+
+	return fmt.Sprintf(
+		"[%v] Downloaded: %.1fMB, Extracted %.1fMB",
+		bar,
+		float64(downloaded)/1e6,
+		float64(extracted)/1e6,
+	)
+}
+
+func (f *pullLogFormatter) backoffOrRenderProgress() *string {
+	// log at most one line every 1 second
+	now := time.Now()
+	if now.Before(f.Backoff) {
+		return nil
+	}
+	f.Backoff = now.Add(1 * time.Second)
+
+	progress := f.RenderProgress()
+	return &progress
+}
+
+// Update returns nil or a rendered progress update for the end user.
+func (f *pullLogFormatter) Update(msg jsonmessage.JSONMessage) *string {
+	if msg.Error != nil {
+		fmt.Printf("%v\n", msg.Error)
+		return nil
+	}
+
+	if msg.Status == "Pulling fs layer" || msg.Status == "Waiting" {
+		if _, ok := f.Known[msg.ID]; !ok {
+			// New layer!
+			f.Known[msg.ID] = &pullInfo{}
+			f.Order = append(f.Order, msg.ID)
+		}
+		return nil
+	}
+
+	if msg.Status == "Downloading" {
+		info := f.Known[msg.ID]
```

Comment:
check existence and error log if missing but don't just panic (the access below will deref and panic)?